File: leetcode0962/solutoin.go
/*
Solution: Monotonic Stack

Time: O(n) | Space: O(n)

Runtime: 48 ms | 31.58%
Memory: 7.22 MB | 47.37%

- n: length of 'nums'
*/

func maxWidthRamp(nums []int) int {
    n := len(nums);

    stack := []int {};

    res := 0;

    for i := 0; i < n; i++ {
        if len(stack) == 0 || nums[stack[len(stack) - 1]] > nums[i] {
            stack = append(stack, i);
        }
    }

    for i:= n - 1; i >= 0; i-- {
        for len(stack) > 0 {
            left := stack[len(stack) - 1];
            if nums[left] <= nums[i] {
                if res < i - left {
                    res = i - left;
                }
                stack = stack[:len(stack) - 1];
            } else {
                break;
            }
        }
    }

    return res;
}

File: leetcode3264/solutoin.rs
/*
Solution: Binary Heap

Time: O(n log n) | Space: O(n)

Runtime: 0 ms | 100.00%
Memory: 2.40 MB | 20.00%

- n: length of 'nums'
*/

use std::collections::BinaryHeap;

impl Solution {
    pub fn get_final_state(nums: Vec<i32>, k: i32, multiplier: i32) -> Vec<i32> {
        let n: usize = nums.len();

        let mut heap: BinaryHeap<(i32, i32)> = BinaryHeap::new();

        for i in 0..nums.len() {
            heap.push((-nums[i], -(i as i32)));
        }

        for _ in 0..k {
            if let Some((num, idx)) = heap.pop() {
                heap.push((multiplier * num, idx));
            }
        }

        let mut res: Vec<i32> = vec![0; n];

        while let Some((num, idx)) = heap.pop() {
            let num: i32 = -num;
            let idx: usize = -idx as usize;

            res[idx] = num;
        }

        res
    }
}